common/src/main/java/io/pravega/common/tracing/RequestTracker.java

```diff
@@ -0,0 +1,188 @@
+
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * This class is a helper for multiple classes to get the client-generated id for a request (requestId) based on the
+ * available information of the request itself (requestDescriptor) in multiple methods. The objective is to cache such
+ * [requestDescriptor, requestId] pairs to we allow classes within a component (Controller, Segment Store) to also log
+ * the client-generated ids, making it easier to trace the lifecycle of a request across multiple logs.
+ */
+@Slf4j
+public final class RequestTracker {
+
+    private static final String INTER_FIELD_DELIMITER = "-";
+    private static final int MAX_CACHE_SIZE = 1000000;
+    private static final int EVICTION_PERIOD_MINUTES = 10;
+
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    private final Cache<String, List<Long>> ongoingRequests;
+
+    public RequestTracker() {
+        // Clean request tags after a certain amount of time.
+        ongoingRequests = CacheBuilder.newBuilder()
+                                      .maximumSize(MAX_CACHE_SIZE)
+                                      .expireAfterWrite(EVICTION_PERIOD_MINUTES, TimeUnit.MINUTES)
+                                      .build();
+    }
+
+    /**
+     * Creates a request descriptor or key to locate the client request id.
+     *
+     * @param requestInfo Fields to form the request descriptor.
+     * @return Request descriptor.
+     */
+    public static String buildRequestDescriptor(String...requestInfo) {
+        return Stream.of(requestInfo).collect(Collectors.joining(INTER_FIELD_DELIMITER));
+    }
+
+    /**
+     * Retrieves a RequestTag object formed by a request descriptor and request id pair. In the case of concurrent
+     * requests with the same descriptor,
+     *
+     * @param requestInfo Fields to form the request descriptor.
+     * @return Request descriptor and request id pair embedded in a RequestTag object.
+     */
+    public RequestTag getRequestTagFor(String...requestInfo) {
+        return getRequestTagFor(RequestTracker.buildRequestDescriptor(requestInfo));
+    }
+
+    public RequestTag getRequestTagFor(String requestDescriptor) {
+        Preconditions.checkNotNull(requestDescriptor, "Attempting to get a null request descriptor.");
+        long requestId;
+        List<Long> descriptorIds;
+
+        synchronized (lock) {
+            descriptorIds = ongoingRequests.getIfPresent(requestDescriptor);
+            requestId = (descriptorIds == null || descriptorIds.size() == 0) ? RequestTag.NON_EXISTENT_ID : descriptorIds.get(0);
```

Comment:
Ok, this is odd, especially because there is no javadoc talking about this and stating the precise properties this method is providing. I'm not sure why this expects multiple descriptor IDs, and why it would get the first of the list. I suppose that if we get multiple, then any is good and we pick the first for simplicity? 

The PDP has some explanation for this, but we better add at least some of that explanation to the javadocs to avoid confusion. Pointing to the PDP should also be ok.

Reply:
Javadoc included and PDP referenced. Yes, the idea is to have a consistent position to track concurrent requests with the same request descriptor. This will allow us to at least track the request that will do the actual changes (others will get rejected, as traced operations are idempotent).